core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -471,25 +471,33 @@ public void close() {
    * @param workerNetAddress the network address of the channel
    * @return the acquired block worker
    */
-  public BlockWorkerClient acquireBlockWorkerClient(final WorkerNetAddress workerNetAddress)
+  public CloseableResource<BlockWorkerClient> acquireBlockWorkerClient(
+      final WorkerNetAddress workerNetAddress)
       throws IOException {
     try (ReinitBlockerResource r = blockReinit()) {
-      return acquireBlockWorkerClientInternal(workerNetAddress, getClientContext().getSubject());
+      return acquireBlockWorkerClientInternal(workerNetAddress, getClientContext());
     }
   }
 
-  private BlockWorkerClient acquireBlockWorkerClientInternal(
-      final WorkerNetAddress workerNetAddress, final Subject subject) throws IOException {
-    SocketAddress address =
-        NetworkAddressUtils.getDataPortSocketAddress(workerNetAddress, getClusterConf());
+  private CloseableResource<BlockWorkerClient> acquireBlockWorkerClientInternal(
+      final WorkerNetAddress workerNetAddress, final ClientContext context) throws IOException {
+    SocketAddress address = NetworkAddressUtils
+        .getDataPortSocketAddress(workerNetAddress, context.getClusterConf());
     GrpcServerAddress serverAddress = GrpcServerAddress.create(workerNetAddress.getHost(), address);
-    ClientPoolKey key = new ClientPoolKey(address,
-        AuthenticationUserUtils.getImpersonationUser(subject, getClusterConf()));
-    return mBlockWorkerClientPool.computeIfAbsent(key,
-        k -> new BlockWorkerClientPool(subject, serverAddress,
-            getClusterConf().getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_SIZE),
-            getClusterConf(), mWorkerGroup))
-        .acquire();
+    ClientPoolKey key = new ClientPoolKey(address, AuthenticationUserUtils
+            .getImpersonationUser(context.getSubject(), context.getClusterConf()));
+    return new CloseableResource<BlockWorkerClient>(mBlockWorkerClientPool.computeIfAbsent(key,
+        k -> new BlockWorkerClientPool(context.getSubject(), serverAddress,
+            context.getClusterConf().getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_SIZE),
+            context.getClusterConf(), mWorkerGroup))
+        .acquire()) {
+      // Save the reference to the original pool map.
+      ConcurrentHashMap<ClientPoolKey, BlockWorkerClientPool> mPoolMap = mBlockWorkerClientPool;
```

Comment:
Hrmmm, can we avoid explicitly defining it, and just define a local final var for it? It will get captured like that right?

Reply:
done